markovflow/sde/sde_utils.py

```diff
@@ -0,0 +1,154 @@
+    n_batch = x0.shape[0]
+
+    dt = tf.convert_to_tensor([[time_interval[1] - time_interval[0]]], dtype=DTYPE)
+    f = sde.drift
+    l = sde.diffusion
+
+    def _step(current_val, nxt_val):
+        x, t = current_val
+        _, t_nxt = nxt_val
+        diffusion_term = l(x, t) * tf.math.sqrt(dt)
+        x_nxt = x + f(x, t) * dt + tf.random.normal(x.shape, dtype=DTYPE) @ diffusion_term
+        return x_nxt, t_nxt
+
+    # [num_data, batch_shape, state_dim] for tf.scan
+    sde_values = tf.zeros((num_data, n_batch, state_dim), dtype=DTYPE)
+
+    # Adding time for batches for tf.scan
+    t0 = tf.zeros((n_batch, 1), dtype=DTYPE)
+    time_interval = tf.reshape(time_interval, (-1, 1, 1))
+    time_interval = tf.repeat(time_interval, n_batch, axis=1)
+
+    sde_values, _ = tf.scan(_step, (sde_values, time_interval), (x0, t0))
+
+    # [batch_shape, num_data, state_dim]
+    sde_values = tf.transpose(sde_values, [1, 0, 2])
+
+    # Appending the initial value
+    sde_values = tf.concat([tf.expand_dims(x0, axis=1), sde_values], axis=1)
+
+    return sde_values
+
+
+def E_sde_drift(sde: SDE, q_mean: TensorType, q_covar: TensorType) -> TensorType:
+    """
+    Calculates the Expectation of the SDE's drift under the provided Gaussian over states.
+
+    ..math:: E_q(x(t))[f(x(t))]
+
+    :param sde: SDE to be linearized.
+    :param q_mean: mean of Gaussian over states with shape (n_batch, num_states, state_dim).
+    :param q_covar: covariance of Gaussian over states with shape (n_batch, num_states, state_dim, state_dim).
+
+    :return: the expectation value with shape (n_batch, num_states, state_dim).
+    """
+    fx = lambda x: sde.drift(x=x, t=tf.zeros(x.shape[0], 1))
+
+    n_batch, n_states, state_dim = q_mean.shape
+    q_mean = tf.reshape(q_mean, (-1, state_dim))
+    q_covar = tf.reshape(q_covar, (-1, state_dim, state_dim))
+
+    val = mvnquad(fx, q_mean, q_covar, H=10)
+
+    val = tf.reshape(val, (n_batch, n_states, state_dim))
+    return val
+
+
+def E_sde_drift_gradient(sde: SDE, q_mean: TensorType, q_covar: TensorType) -> TensorType:
+    """
+     Calculates the Expectation of the gradient of the SDE's drift under the provided Gaussian over states
+
+    ..math:: E_q(.)[f'(x(t))]
+
+    :param sde: SDE to be linearized.
+    :param q_mean: mean of Gaussian over states with shape (n_batch, num_states, state_dim).
+    :param q_covar: covariance of Gaussian over states with shape (n_batch, num_states, state_dim, state_dim).
+
+    :return: the expectation value with shape (n_batch, num_states, state_dim).
+    """
+    n_batch, n_states, state_dim = q_mean.shape
+    q_mean = tf.reshape(q_mean, (-1, state_dim))
+    q_covar = tf.reshape(q_covar, (-1, state_dim, state_dim))
+    val = mvnquad(sde.sde_drift_gradient, q_mean, q_covar, H=10)
+
+    val = tf.reshape(val, (n_batch, n_states, state_dim))
+    return val
+
+
+def linearize_sde(sde: SDE, q_mean: TensorType, q_covar: TensorType, initial_mean: TensorType,
```

Comment:
you will need the time grid as input (actually the dt but passing the whole time_grid is better I'd say)